src/dbnode/encoding/proto/encoder.go

```diff
@@ -0,0 +1,807 @@
+	neg := false
+	enc.customFields[i].prevFloatBits = uint64(v)
+	if v < 0 {
+		neg = true
+		v = -1 * v
+	}
+
+	vBits := uint64(v)
+	numSig := encoding.NumSig(vBits)
+
+	m3tsz.WriteIntSig(enc.stream, &enc.customFields[i].intSigBitsTracker, numSig)
+	enc.encodeIntValDiff(vBits, neg, numSig)
+}
+
+func (enc *Encoder) encodeFirstUnsignedIntValue(i int, v uint64) {
+	enc.customFields[i].prevFloatBits = uint64(v)
+
+	vBits := uint64(v)
+	numSig := encoding.NumSig(vBits)
+
+	m3tsz.WriteIntSig(enc.stream, &enc.customFields[i].intSigBitsTracker, numSig)
+	enc.encodeIntValDiff(vBits, false, numSig)
+}
+
+func (enc *Encoder) encodeNextSignedIntValue(i int, next int64) {
+	prev := int64(enc.customFields[i].prevFloatBits)
+	diff := next - prev
+	if diff == 0 {
+		enc.stream.WriteBit(opCodeNoChange)
+		return
+	}
+
+	enc.stream.WriteBit(opCodeChange)
+
+	neg := false
+	if diff < 0 {
+		neg = true
+		diff = -1 * diff
+	}
+
+	var (
+		diffBits = uint64(diff)
+		numSig   = encoding.NumSig(diffBits)
+		newSig   = enc.customFields[i].intSigBitsTracker.TrackNewSig(numSig)
+	)
+	// Can't use an intermediary variable for the intSigBitsTracker because we need to
+	// modify the value in the slice, not a copy of it.
+	m3tsz.WriteIntSig(enc.stream, &enc.customFields[i].intSigBitsTracker, newSig)
+	enc.encodeIntValDiff(diffBits, neg, newSig)
+	enc.customFields[i].prevFloatBits = uint64(next)
+}
+
+func (enc *Encoder) encodeNextUnsignedIntValue(i int, next uint64) {
+	var (
+		neg  = false
+		prev = uint64(enc.customFields[i].prevFloatBits)
+		diff uint64
+	)
+
+	// Avoid overflows.
+	if next > prev {
+		diff = next - prev
+	} else {
+		neg = true
+		diff = prev - next
+	}
+
+	if diff == 0 {
+		enc.stream.WriteBit(opCodeNoChange)
+		return
+	}
+
+	enc.stream.WriteBit(opCodeChange)
+
+	numSig := encoding.NumSig(diff)
+	newSig := enc.customFields[i].intSigBitsTracker.TrackNewSig(numSig)
+	// Can't use an intermediary variable for the intSigBitsTracker because we need to
+	// modify the value in the slice, not a copy of it.
+	m3tsz.WriteIntSig(enc.stream, &enc.customFields[i].intSigBitsTracker, newSig)
+	enc.encodeIntValDiff(diff, neg, newSig)
+	enc.customFields[i].prevFloatBits = uint64(next)
+}
+
+func (enc *Encoder) encodeIntValDiff(valBits uint64, neg bool, numSig uint8) {
+	if neg {
+		// opCodeNegative
+		enc.stream.WriteBit(opCodeIntDeltaNegative)
+	} else {
+		// opCodePositive
+		enc.stream.WriteBit(opCodeIntDeltaPositive)
+	}
+
+	enc.stream.WriteBits(valBits, int(numSig))
+}
+
+func (enc *Encoder) moveToEndOfBytesDict(fieldIdx, i int) {
+	existing := enc.customFields[fieldIdx].bytesFieldDict
+	for j := i; j < len(existing); j++ {
+		nextIdx := j + 1
+		if nextIdx >= len(existing) {
+			break
+		}
+
+		currVal := existing[j]
+		nextVal := existing[nextIdx]
+		existing[j] = nextVal
+		existing[nextIdx] = currVal
+	}
+}
+
+func (enc *Encoder) addToBytesDict(fieldIdx int, hash uint64) {
+	existing := enc.customFields[fieldIdx].bytesFieldDict
```

Comment:
You can't implement an LRU with a heap can you? Would need a map + double linked list

Reply:
Current algo shouldn't have repeated hashes unless there is a hash collision.

Doubly linked list would work, gonna leave it out for now though since these shouldn't be too big. Can add as a follow-up

Reply:
Yeah I think this should be efficient for smaller numbers - just suggesting having a bench to see where the breaking point is of when we should switch it out (eg if the benchmarks show that it won't matter until we get to thousands, then we'll never need to do it, but if it shows that it'll be way more efficient at 20, we may want to add a TODO).

Reply:
Yeah I see what you're saying. I'm gonna leave it out for now because we're probably gonna want to make the dictionary compression more sophisticated in the future and we'll need a datastructure like a tree instead of the array LRU anyways